builtin/logical/pki/path_fetch_keys.go

```diff
@@ -0,0 +1,225 @@
+or its assigned name value.
+
+Writing to /key/:ref allows updating of the name field associated with
+the certificate.
+`
+)
+
+func (b *backend) pathGetKeyHandler(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	keyRef := data.Get("ref").(string)
+	if len(keyRef) == 0 {
+		return logical.ErrorResponse("missing key reference"), nil
+	}
+
+	keyId, err := resolveKeyReference(ctx, req.Storage, keyRef)
+	if err != nil {
+		return nil, err
+	}
+	if keyId == "" {
+		return logical.ErrorResponse("unable to resolve key id for reference" + keyRef), nil
+	}
+
+	key, err := fetchKeyById(ctx, req.Storage, keyId)
+	if err != nil {
+		return nil, err
+	}
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			keyIdParam:   key.ID,
+			keyNameParam: key.Name,
+			keyTypeParam: key.PrivateKeyType,
+		},
+	}, nil
+}
+
+func (b *backend) pathUpdateKeyHandler(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	keyRef := data.Get(keyRefParam).(string)
+	if len(keyRef) == 0 {
+		return logical.ErrorResponse("missing key reference"), nil
+	}
+
+	keyId, err := resolveKeyReference(ctx, req.Storage, keyRef)
+	if err != nil {
+		return nil, err
+	}
+	if keyId == "" {
+		return logical.ErrorResponse("unable to resolve key id for reference" + keyRef), nil
+	}
+
+	key, err := fetchKeyById(ctx, req.Storage, keyId)
+	if err != nil {
+		return nil, err
+	}
+
+	newName := data.Get(keyNameParam).(string)
+	if len(newName) > 0 && !nameMatcher.MatchString(newName) {
+		return logical.ErrorResponse("new key name outside of valid character limits"), nil
+	}
+
+	if newName != key.Name {
+		key.Name = newName
+
+		err := writeKey(ctx, req.Storage, *key)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	resp := &logical.Response{
+		Data: map[string]interface{}{
+			keyIdParam:   key.ID,
+			keyNameParam: key.Name,
+			keyTypeParam: key.PrivateKeyType,
+		},
+	}
+
+	if len(newName) == 0 {
+		resp.AddWarning("Name successfully deleted, you will now need to reference this key by it's Id: " + string(key.ID))
+	}
+
+	return resp, nil
+}
+
+func (b *backend) pathDeleteKeyHandler(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	keyRef := data.Get(keyRefParam).(string)
+	if len(keyRef) == 0 {
+		return logical.ErrorResponse("missing key reference"), nil
+	}
+
+	keyId, err := resolveKeyReference(ctx, req.Storage, keyRef)
+	if err != nil {
+		return nil, err
+	}
+	if keyId == "" {
+		return logical.ErrorResponse("unable to resolve key id for reference" + keyRef), nil
+	}
+
+	keyInUse, issuerId, err := isKeyInUse(keyId.String(), ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+	if keyInUse {
+		return logical.ErrorResponse(fmt.Sprintf("Failed to Delete Key.  Key in Use by Issuer: %s", issuerId)), nil
+	}
+
+	wasDefault, err := deleteKey(ctx, req.Storage, keyId)
```

Comment:
Not your fault at all but you can get a scenario like this:

```
 - import or generate stuff
 - delete a key associated with an issuer
 - do something else
```

and end up with a dangling ref:

```
* 1 error occurred:
	* error building CRLs: unable to build CRL for issuer (f7eeeab2-9c58-58f6-612a-43b280c4e308): could not fetch the CA certificate: pki key id a051f23f-e12e-46d2-dc20-2e9c47f9963c does not exist
```

(that key ID doesn't exist)

Do you want to take on fixing up `deleteKey` to make sure we update issuer's KeyID fields if they point to this keyId that's being deleted?

Reply:
Sorry, I thought that was fixed here: https://github.com/hashicorp/vault/pull/15150/commits/00dd62575a3de59e80407d51696dbc132989f69b (like eight lines up).  
Are you able to reproduce this?

It's totally doable to guard against this within deleteKey instead, i just think it's kind of ugly.  If we wanted to, for instance, delete the endpoint by deleting all issuers, and all keys, I wouldn't want to be making a call against issuers for each key deleted (though presumably we'd have deleted issuers first).  I'd also need to change the signature of deleteKey to return an issuerId if it existed.  Let me know if that's preferred though.

Reply:
Ah, interesting. Lets me try again, I definitely thought I had that...

Reply:
Woah, magic, thank you.  This is really helpful.